File: android/app/src/main/java/com/nextpage/data/remote/catalog/GoogleBooksDataSource.kt
package com.nextpage.data.remote.catalog

import java.net.URLEncoder
import kotlinx.serialization.json.Json

const val GOOGLE_BOOKS_BASE_URL = "https://www.googleapis.com/books/v1"

/**
 * Google Books datasource (U2) — keyed volume search by ISBN / title-author.
 *
 * Fail-closed on the key: the provider is only built when
 * `GOOGLE_BOOKS_KEY` (gitignored `local.properties` → `BuildConfig`) is
 * non-blank (see `googleBooksProviderOrNull`). This datasource itself takes
 * the key as a constructor arg so it stays fully testable offline with
 * [FakeCatalogHttpTransport]; the key is appended as `&key=` only when
 * non-blank and never logged.
 */
open class GoogleBooksDataSource(
    private val transport: CatalogHttpTransport,
    private val apiKey: String = "",
    private val baseUrl: String = GOOGLE_BOOKS_BASE_URL,
    private val json: Json = Json { ignoreUnknownKeys = true }
) {

    /** Search volumes; ISBN queries use the `isbn:` operator, all else raw. */
    open suspend fun search(
        query: String,
        page: Int,
        pageSize: Int = DEFAULT_PAGE_SIZE
    ): CatalogSearchResult {
        val size = clampPageSize(pageSize)
        val startIndex = (page - 1) * size
        val q = buildGoogleBooksQuery(query)
        val keyParam = apiKey.trim().takeIf { it.isNotEmpty() }?.let { "&key=${it.encode()}" } ?: ""
        val url = "$baseUrl/volumes?q=${q.encode()}&startIndex=$startIndex&maxResults=$size$keyParam"
        try {
            val res = transport.getWithRetry(url)
            val data = json.decodeFromString<GoogleBooksSearchResponse>(res.body)
            val books = data.items.mapNotNull(::mapGoogleBooksVolume).take(size)
            return CatalogSearchResult(books, data.totalItems ?: books.size)
        } catch (err: Throwable) {
            if (isCatalogError(err)) throw err
            throw catalogError(CatalogErrorCode.NETWORK_ERROR, "googlebooks request failed")
        }
    }

    /** Fetch one volume by id; unknown ids surface NOT_FOUND. */
    open suspend fun getById(volumeId: String): CatalogBook {
        val keyParam = apiKey.trim().takeIf { it.isNotEmpty() }?.let { "&key=${it.encode()}" } ?: ""
        // `fields=` keeps the payload small; unknown fields still decode leniently.
        val url = "$baseUrl/volumes/${volumeId.encode()}?fields=id,volumeInfo(title,authors,description,language,categories,imageLinks)$keyParam"
        try {
            val res = transport.getWithRetry(url)
            val item = json.decodeFromString<GoogleBooksVolumeItem>(res.body)
            return mapGoogleBooksVolume(item)
                ?: throw catalogError(CatalogErrorCode.NOT_FOUND, "googlebooks volume unavailable")
        } catch (err: Throwable) {
            if (isCatalogError(err)) throw err
            throw catalogError(CatalogErrorCode.NETWORK_ERROR, "googlebooks request failed")
        }
    }

    private fun String.encode(): String = URLEncoder.encode(this, Charsets.UTF_8.name())
}

/** ISBN-10 total length, ISBN-10 body length (without check digit), ISBN-13 length. */
private const val ISBN10_LENGTH = 10
private const val ISBN10_BODY_LENGTH = 9
private const val ISBN13_LENGTH = 13

/**
 * Build the `q` param: bare ISBN-10/13 (digits, optional `978`/`979` prefix,
 * `X` check digit, dashes/spaces tolerated) uses the `isbn:` operator so an
 * ISBN search resolves to the exact edition; everything else passes through
 * as a title/author query the API ranks itself.
 */
fun buildGoogleBooksQuery(rawQuery: String): String {
    val trimmed = rawQuery.trim()
    val compact = trimmed.replace(Regex("[-\\s]"), "").uppercase()
    val digits = compact.let { if (it.endsWith("X")) it.dropLast(1) + "X" else it }
    val isIsbn = (digits.length == ISBN10_LENGTH && digits.take(ISBN10_BODY_LENGTH).all { it.isDigit() } &&
        (digits.last().isDigit() || digits.last() == 'X')) ||
        (digits.length == ISBN13_LENGTH && digits.all { it.isDigit() })
    return if (isIsbn) "isbn:$compact" else trimmed
}

File: desktop/src/lib/shared/services/catalog/GoogleBooksDataSource.ts
/**
 * Google Books datasource (WU3) — keyed volume search/detail by ISBN or
 * title-author. Plain `fetch` + identified UA; `connect-src` already
 * allowlists https://www.googleapis.com in tauri.conf.json.
 *
 * Fail-closed on the key: the provider is only constructed when
 * `VITE_GOOGLE_BOOKS_KEY` is non-blank (see `googleBooksProviderOrNull`). The
 * datasource takes the key as a constructor arg so it stays fully testable
 * offline; the key is appended as `&key=` only when non-blank and is never
 * logged, never echoed into an error message.
 */
import { catalogError, isCatalogError } from './errors';
import { DESKTOP_USER_AGENT, clampPageSize, fetchWithRetry, toCatalogError } from './policy';
import {
  mapGoogleBooksVolume,
  type GoogleBooksSearchResponse,
  type GoogleBooksVolumeItem,
} from './mappers';
import type { CatalogBook } from './CatalogProvider';

export const GOOGLE_BOOKS_BASE_URL = 'https://www.googleapis.com/books/v1';

/**
 * `fields` whitelist for the detail endpoint: keeps the payload small while
 * carrying everything the mapper needs. `industryIdentifiers` is included so a
 * keyed detail can surface ISBN-10/13 enrichment (spec scenario "Keyed
 * enrichment"); unknown/missing fields still map leniently.
 */
export const GOOGLE_BOOKS_VOLUME_FIELDS =
  'id,volumeInfo(title,authors,description,language,categories,imageLinks,industryIdentifiers)';

/** A bare `isbn:`-eligible identifier: 9 digits + digit/X check digit, or 13 digits. */
const ISBN10_RE = /^\d{9}[\dX]$/;
const ISBN13_RE = /^\d{13}$/;

/**
 * Build the `q` param: a bare ISBN-10/13 (digits, optional `X` check digit,
 * dashes/spaces tolerated) uses the `isbn:` operator so an ISBN search resolves
 * to the exact edition; everything else passes through as a title/author query
 * the API ranks itself.
 */
export function buildGoogleBooksQuery(rawQuery: string): string {
  const trimmed = rawQuery.trim();
  const compact = trimmed.replace(/[-\s]/g, '').toUpperCase();
  return ISBN10_RE.test(compact) || ISBN13_RE.test(compact) ? `isbn:${compact}` : trimmed;
}

export class GoogleBooksDataSource {
  constructor(
    private readonly fetchFn: typeof fetch = fetch,
    private readonly apiKey: string = '',
    private readonly userAgent: string = DESKTOP_USER_AGENT,
  ) {}

  /** `&key=…` only when the key is non-blank; the key is never logged. */
  private keyParam(): string {
    const trimmed = this.apiKey.trim();
    return trimmed === '' ? '' : `&key=${encodeURIComponent(trimmed)}`;
  }

  private async getJson(url: string): Promise<unknown> {
    try {
      const res = await fetchWithRetry(
        url,
        { headers: { 'User-Agent': this.userAgent, Accept: 'application/json' } },
        this.fetchFn,
      );
      return await res.json();
    } catch (err) {
      if (isCatalogError(err)) throw err;
      toCatalogError(err, 'googlebooks request failed');
    }
  }

  /** Search volumes; ISBN queries use the `isbn:` operator, all else raw. */
  async search(
    query: string,
    page: number,
    pageSize = 24,
  ): Promise<{
    books: CatalogBook[];
    totalCount: number;
  }> {
    const size = clampPageSize(pageSize);
    const startIndex = (page - 1) * size;
    const q = encodeURIComponent(buildGoogleBooksQuery(query));
    const url = `${GOOGLE_BOOKS_BASE_URL}/volumes?q=${q}&startIndex=${startIndex}&maxResults=${size}${this.keyParam()}`;
    const data = (await this.getJson(url)) as GoogleBooksSearchResponse;
    const books = (data.items ?? [])
      .map(mapGoogleBooksVolume)
      .filter((b): b is CatalogBook => b !== null)
      .slice(0, size);
    return {
      books,
      totalCount: typeof data.totalItems === 'number' ? data.totalItems : books.length,
    };
  }

  /** Fetch one volume by id; unknown ids surface NOT_FOUND. */
  async getById(volumeId: string): Promise<CatalogBook> {
    const url = `${GOOGLE_BOOKS_BASE_URL}/volumes/${encodeURIComponent(volumeId)}?fields=${GOOGLE_BOOKS_VOLUME_FIELDS}${this.keyParam()}`;
    const data = (await this.getJson(url)) as GoogleBooksVolumeItem;
    const book = mapGoogleBooksVolume(data);
    if (!book) throw catalogError('NOT_FOUND', 'googlebooks volume unavailable');
    return book;
  }
}
